### ISSUE TITLE
Initial stream delay across NGINX proxy

### ISSUE THREAD

**Issue by MCBama** [django/channels#639]
Setting up websocket communication using `django-channels` with `NGINX` as a proxy may be causing delay issues with Server-to-Client streams. The Server has to send exactly 24 messages regardless of message speed to the Client across the websocket before the Client begins to receive any data.

- Client connected directly to Django test server streams fine
- Client connected directly to Daphne streams fine
- With NGINX acting as proxy neither Django test server nor Daphne server streaming works until after 24 messages have been sent from Server
- Replaced Django server with Node.js socket server and streaming worked fine across NGINX proxy
- Tested connection to Django server through NGINX with both browser and Node.js wscat connection and stream delay existed on both
- **If we send a message on connection (from Client to Django server) through NGINX proxy, streaming from server begins to function immediately after initial message**

Given the above we believe the problem may exist within `django-channels` as the problem only appears when `django-channels` is integrated under a proxy with NGINX while Node.js client/server websocket architecture functions as expected.

**Comment by andrewgodwin** [comment]
I'm going to need exact steps to replicate including precise versions, OS versions and the entire nginx config to be able to confirm this, and I know of plenty of installations (including my own) where this doesn't happen, so it's likely not a channels issue.

**Comment by MCBama** [comment]
It's possible that it's a Windows only bug if your system is running on Linux?

Windows 7 Professional 64-bit
Python 3.5.2
Django 1.11
Channels 1.1.3
nginx 1.12.0

```
# nginx config file
worker_processes  2;

events {
    worker_connections  1024;
}

http {
    map $http_upgrade $connection_upgrade {
      default upgrade;
      '' close;
    }
    include       mime.types;
    default_type  application/octet-stream;
    sendfile        on;
    keepalive_timeout  0;

upstream django {
  keepalive 100;
  server 127.0.0.1:8001;
}
  server {
    listen 80;
    client_max_body_size 20M;
    server_name localhost;
    tcp_nodelay     on;

    location / {
      proxy_buffering off;
      proxy_pass http://django;
      proxy_read_timeout     300;
      proxy_connect_timeout  300;

      proxy_http_version 1.1;
      proxy_set_header Upgrade $http_upgrade;
      proxy_set_header Connection "upgrade";

      proxy_redirect     off;
      proxy_set_header   Host $host;
      proxy_set_header   X-Real-IP $remote_addr;
      proxy_set_header   X-Forwarded-For $proxy_add_x_forwarded_for;
      proxy_set_header   X-Forwarded-Host $server_name;

    }
  }
}
```
Routing:
```
# in routing.py
from channels.routing import route, include, route_class
from vits.consumers import MissionConsumer
from vits.consumers import FlightConsumer

routing = [
  # build routes for mission sockets
  route_class(MissionConsumer, path=r"^/mission"),
  route_class(FlightConsumer, path=r'^/flight/(?P<pk>[^/]+)'),
  route_class(FlightConsumer, path=r'^/flight/all_flight_updates'),
]
```
Consumers:
```
# In consumers.py
from channels import Channel
from channels.sessions import channel_session
from channels import Group

from channels.generic.websockets import JsonWebsocketConsumer

class MissionConsumer(JsonWebsocketConsumer):
  http_user = True

  def connection_groups(self, **kwargs):
    group_list = []
    if "update" in self.path:
      group_list.append("mission_updated")
    # To filter groups open to socket, you can use
    # self.message.user to get the session user and filter
    # acceptable groups off of the user's groups
    if self.message.user.groups.filter(name="admin").exists():
      pass

    return group_list

  def connect(self, message, **kwargs):
    super(MissionConsumer, self).connect(message, **kwargs)


  def receive(self, content, **kwargs):
    # Stick the message onto the processing queue
    super(MissionConsumer, self).receive(content, **kwargs)


class FlightConsumer(JsonWebsocketConsumer):
  http_user = True

  def connection_groups(self, **kwargs):
    group_list = []
    if "/flight_pos" in self.path:
      group_list.append("flight_{}_new_position".format(kwargs['pk']))
    elif "/flight_events" in self.path:
      group_list.append("flight_{}_event_created".format(kwargs['pk']))
    elif "/flight_updates" in self.path:
      group_list.append("flight_{}_updated".format(kwargs['pk']))
    elif '/all_flight_updates' in self.path:
      group_list.append("flight_updated")
    elif '/flight_created' in self.path:
      group_list.append("flight_created")
    return group_list

  def connect(self, message, **kwargs):
    super(FlightConsumer, self).connect(message, **kwargs)

  def receive(self, content, **kwargs):
    # Stick the message onto the processing queue
    print(content)
    self.send(content)
```

Client:
```
var flights_socket = new WebSocket("ws://"+window.location.host + "/flight/all_flight_updates");
  message_count = 0;
  flights_socket.onmessage = function(e) {
    var flight = JSON.parse(e.data)[0];
    if("undefined" === typeof(flight.text))
    {
      console.log(flight)
      var new_point = {
        'lat': flight.fields.latitude,
        'lon': flight.fields.longitude
      }
      if(history_loaded){
        updateFlightMap(flight.pk.toString(), new_point);
      }
    }
  }
    flights_socket.onopen = function() {
      console.log("Listening for flight updates");
      //flights_socket.send(JSON.stringify([{'text':'connected'}]));
  }
  // Call onopen directly if socket is already open
  if (flights_socket.readyState == WebSocket.OPEN) flight_socket.onopen();
```

**Comment by MCBama** [comment]
The only steps to reproduce I can really supply are thus:

1. Connect client to server across any channel
2. Ensure client never sends any form of message to the server 
3. Start streaming data from server to client
4. 24 data streams later, client will receive all 24 buffered data streams and will continue receiving data in real time from this point forward.

To remove delay:
1. Connect client to server across any channel
2. Send any message from client to server upon connection (any message seems to work)
3. Start streaming data from server to client
4. Client will immediately begin receiving data traffic without delay.

**Comment by andrewgodwin** [comment]
I develop on Windows 10, so that is unlikely. I will look into this at some point, but I'll have to triage it as low-priority until I get a second complaint or get spare time to verify.

**Comment by MCBama** [comment]
Understandable. I'm in the process of writing a quick and dirty example app that I should be able to pull on different computers to test to see if it is environment specific or if it is reproducible. If it's helpful I'll share that with you once it's finished so with any luck all you'd have to do is pull the app down and give it a run.

**Comment by MCBama** [comment]
**Update:**

- Realized I was running Python 3.5.2 32 bit on a 64 bit machine
- Updated machine to Pythion 3.6.1 64 bit
- Server now sends data real-time from the beginning with no delay through NGINX without initial message

So...either changing from 32 bit Python to 64 bit Python or upgrading from Python 3.5.2 to Python 3.6.1 solved the issue. I am unsure as to which however.

Thanks for your time responding to my request. If you decide you want to delve deeper into this to see if it can be reproduced elsewhere I'll be happy to uninstall 3.6.1 and install 3.5.2 64 bit to see if that has any change to assist tracking down what happened.

**Comment by MCBama** [comment]
**Yet another Update**

You probably already received an email update about my last update...but it turns out the only other change I made in addition the the version update was to change the NGINX listen port from default port 80 to port 8000 and out of curiosity I reverted that change and found out that it was actually the port change that had fixed the problem.

Thus...this is probably not your problem after all. I apologize for wasting your time chasing a wild goose.
